src/main/kotlin/build/buf/intellij/resolve/BufRootsProvider.kt

```diff
@@ -72,32 +86,56 @@ class BufRootsProvider : AdditionalLibraryRootsProvider() {
             ?.findFileByRelativePath("${mod.remote}/${mod.owner}/${mod.repository}/${mod.commit}")
 
         fun getOrCreateModuleCacheFolderV2(mod: BufModuleCoordinates, env: Map<String, String> = System.getenv()): VirtualFile? {
-            val bufPath = Paths.get(PathManager.getSystemPath(), "buf")
+            val bufPath = getIDEAModCacheV2()
             if (!bufPath.isDirectory() && !bufPath.toFile().mkdirs()) {
+                LOG.warn("failed to create directory: $bufPath")
                 return null
             }
             val commitPath = bufPath.resolve("${mod.remote}/${mod.owner}/${mod.repository}/${mod.commit}")
             if (commitPath.isDirectory()) {
                 return LocalFileSystem.getInstance().findFileByNioFile(commitPath)
             }
+            if (!commitPath.parent.isDirectory() && !commitPath.parent.toFile().mkdirs()) {
+                LOG.warn("failed to create directory: ${commitPath.parent}")
+                return null
+            }
             val manifest = findManifestV2(mod, env) ?: return null
-            for (path in manifest.getPaths()) {
-                val canonicalPath = manifest.getCanonicalPath(path) ?: return null
-                val modulePath = commitPath.resolve(path)
-                if (!modulePath.parent.isDirectory()) {
-                    Files.createDirectories(modulePath.parent)
+            val commitTempPath = Files.createTempDirectory(commitPath.parent, "tmp_" + mod.commit)
+            try {
+                for (path in manifest.getPaths()) {
+                    val canonicalPath = manifest.getCanonicalPath(path) ?: return null
+                    val modulePath = commitTempPath.resolve(path)
+                    if (!modulePath.parent.isDirectory()) {
+                        Files.createDirectories(modulePath.parent)
+                    }
+                    Files.copy(Path.of(canonicalPath), modulePath)
+                }
+                try {
+                    Files.move(commitTempPath, commitPath, StandardCopyOption.ATOMIC_MOVE)
```

Comment:
This will perform an atomic rename to the proper place when done (so we avoid race conditions). If someone else won the race, we'll fall into the catch block below and clean things up in the finally block.